src/transport_ofi.c

```diff
@@ -598,52 +598,59 @@ static inline int atomic_limitations_check(void)
     return 0;
 }
 
-static inline int exchange_and_av_insert(struct fabric_info *info)
+static inline int publish_av_info(struct fabric_info *info)
 {
-    int                 i, rank = 0, ret = 0;
-    char   epname[128], *alladdrs = NULL;
+    int    ret = 0;
+    char   epname[128];
     size_t epnamelen = sizeof(epname);
 
-    rank = shmem_runtime_get_rank();
-
     ret = fi_getname((fid_t)shmem_transport_ofi_epfd, epname, &epnamelen);
     if(ret!=0 || (epnamelen > sizeof(epname))){
         OFI_ERRMSG("fi_getname failed\n");
         return ret;
     }
 
-    alladdrs = malloc(info->npes * epnamelen);
-    if(alladdrs==NULL){
-        OFI_ERRMSG("alladdrs is NULL\n");
+    ret = shmem_runtime_put("fi_epname", epname, epnamelen);
+    if (ret != 0) {
+        OFI_ERRMSG("shmem_runtime_put failed\n");
         return ret;
     }
 
-    ret = shmem_runtime_put("OFI", epname, epnamelen);
+    ret = shmem_runtime_put("fi_epnamelen", &epnamelen, sizeof(size_t));
     if (ret!=0) {
         OFI_ERRMSG("shmem_runtime_put failed\n");
         return ret;
     }
 
-    shmem_runtime_exchange();
-    shmem_runtime_barrier();
+    return ret;
+}
 
-    for(i=0; i<info->npes; i++)
-    {
-	    void *tgt = alladdrs + i*epnamelen;
-	    if (i == rank) {
-		    memcpy(tgt, epname, epnamelen);
-	    } else {
-		    shmem_runtime_get(i,"OFI", tgt, epnamelen);
-	    }
+static inline int populate_av()
+{
+    int    i, ret = 0;
+    char   *alladdrs = NULL, *addr_ptr;
+    size_t max_epnamelen = 128;
+
+    alladdrs = malloc(shmem_internal_num_pes * max_epnamelen);
+    if (alladdrs == NULL) {
+        OFI_ERRMSG("Out of memory allocating 'alladdrs'\n");
+        return ret;
+    }
+
+    for (i = 0, addr_ptr = alladdrs; i < shmem_internal_num_pes; i++) {
+            size_t epnamelen;
+            shmem_runtime_get(i, "fi_epnamelen", &epnamelen, sizeof(size_t));
+            shmem_runtime_get(i, "fi_epname", addr_ptr, epnamelen);
```

Comment:
As we discussed in the chat, you only need the epnamelen from PMI if you're going to check that you're not in a heterogeneous fabric. Otherwise, if you assume that the user has chosen the same fabric, same provider on all PEs, you don't need to get this from every PE. You can just assume that all PE name length == local PE name length.

If you are getting the name length from every PE, then at least we can check if it is the same as local name length as a caution - although the best thing to exchange and check would be info->addr_format, since you could have two PEs (in an heterogenous case) with same sized address, but different addr_format.

My suggestion is to be lazy assume epnamelen is same across the job, and its the users job to ensure that they don't break that assumption.